backend/server/src/derivation.rs

```diff
@@ -0,0 +1,142 @@
+//! Derivation build data model.
+
+use std::num::NonZero;
+
+/// A derivation id combined with a counter that keeps track of the number of build attempts for
+/// this derivation.
+pub struct DrvBuildId {
+    /// The derivation that is attempted to be build.
+    pub derivation: DrvId,
+
+    /// Denotes the number of times this derivation has been attempted to be build.
+    ///
+    /// Note that once the build outcome of a derivation has been determined, there is no point in
+    /// trying to build the same derivation again. If it failed once, it will always fail.
+    ///
+    /// This counter is intended for cases in which the derivation build was interrupted due to
+    /// external factors (see [DrvBuildState::Interrupted]). In these situations it may make sense
+    /// to reattempt the build (depending on the interruption kind).
+    pub build_attempt: NonZero<u32>,
+}
+
+/// Metadata about a derivation build.
+///
+/// This metadata is useful to reproduce a build on a different machine. Note that in general,
+/// only builds that ended with a state of [DrvBuildState::Completed] can be reproduced.
+pub struct DrvBuildMetadata {
+    /// The derivation build this metadata is associated with.
+    pub build: DrvBuildId,
+
+    /// The Git commit this derivation build is based upon.
+    ///
+    /// Note that this may not be the only commit that can produce this derivation. Because a
+    /// derivation only needs to fully build once, later commits may still include this
+    /// derivation but do not trigger a new build.
+    pub commit: gix_hash::ObjectId,
+
+    /// The Nix command that was used to build this derivation.
+    pub build_command: DrvBuildCommand,
+}
+
+/// Command used to build the derivation.
+pub enum DrvBuildCommand {
+    /// Nix 1.0 style command. Fields are TBD.
+    Nix1,
+}
```

Comment:
Eventually, I was thinking of allowing different "evaluation modes". But for now we can have something which has:

- Contains the nix packaging logic:
  - repo
  - commit
- Selects the right attr (similar to "legacy" hydra jobsets):
  - file
  - nix args (optional)
  - attr (e.g. `python3Packages.protobuf`)

Reply:
Yes makes sense. I think the Git repo will probably always be necessary, independent from the evaluation mode right?

Actually, does that mean that we should store both an evaluation command and a build command? Or is it enough to have one enum that denotes the "mode" and has all the information requited for both eval and build?

Complexity of this struct does not really matter btw, I plan to just store it json encoded in a single database column.

Reply:
I expanded the `DrvBuildCommand`.